auth/api/iam/openid4vp.go

```diff
@@ -286,6 +309,253 @@ func (r Wrapper) sendAndHandleDirectPostError(ctx context.Context, auth2Error oa
+		codeParam:  authorizationCode,
+		stateParam: oauthSession.ClientState,
+	})
+	return HandleAuthorizeResponse200JSONResponse{RedirectURI: redirectURI.String()}, nil
+}
+
+func withCallbackURI(err error, callbackURI *url.URL) error {
+	oauthErr := err.(oauth.OAuth2Error)
+	oauthErr.RedirectURI = callbackURI
+	return oauthErr
+}
+
+// extractChallenge extracts the nonce from the presentation.
+// it uses the nonce from the JWT if available, otherwise it uses the challenge from the LD proof.
+func extractChallenge(presentation vc.VerifiablePresentation) (string, error) {
+	var nonce string
+	switch presentation.Format() {
+	case vc.JWTPresentationProofFormat:
+		nonceRaw, _ := presentation.JWT().Get("nonce")
+		nonce, _ = nonceRaw.(string)
+	case vc.JSONLDPresentationProofFormat:
+		proof, err := credential.ParseLDProof(presentation)
+		if err != nil {
+			return "", err
+		}
+		if proof.Challenge != nil && *proof.Challenge != "" {
+			nonce = *proof.Challenge
+		}
+	}
+	return nonce, nil
+}
+
+// validatePresentationNonce checks if the nonce is the same for all presentations.
+// it deletes all nonces from the session store in the process.
+// errors are returned as OAuth2 errors.
+func (r Wrapper) validatePresentationNonce(presentations []vc.VerifiablePresentation) error {
+	var nonce string
+	var returnErr error
+	for _, presentation := range presentations {
+		nextNonce, err := extractChallenge(presentation)
+		_ = r.oauthNonceStore().Delete(nextNonce)
+		if nextNonce == "" {
+			// fallback on nonce instead of challenge, todo: should be uniform, check vc data model specs for JWT/JSON-LD
+			nextNonce, err = extractNonce(presentation)
+			if nextNonce == "" {
+				// error when all presentations are missing nonce's
+				returnErr = oauth.OAuth2Error{
+					Code:          oauth.InvalidRequest,
+					InternalError: err,
+					Description:   "presentation has invalid/missing nonce",
+				}
+			}
+		}
+		if nonce != "" && nonce != nextNonce {
+			returnErr = oauth.OAuth2Error{
+				Code:        oauth.InvalidRequest,
+				Description: "not all presentations have the same nonce",
+			}
+		}
+		nonce = nextNonce
+	}
+
+	return returnErr
+}
+
+func (r Wrapper) handleAccessTokenRequest(ctx context.Context, verifier did.DID, authorizationCode *string, redirectURI *string, clientId *string) (HandleTokenRequestResponseObject, error) {
+	// first check redirectURI
+	if redirectURI == nil {
+		return nil, oauthError(oauth.InvalidRequest, "missing redirect_uri parameter")
+	}
+	callbackURI, err := url.Parse(*redirectURI)
+	if err != nil {
+		return nil, oauthError(oauth.InvalidRequest, "invalid redirect_uri parameter")
+	}
+
+	// check if the authorization code is valid
+	var oauthSession OAuthSession
+	err = r.oauthCodeStore().Get(*authorizationCode, &oauthSession)
+	if err != nil {
+		return nil, withCallbackURI(oauthError(oauth.InvalidRequest, "invalid authorization code"), callbackURI)
```

Comment:
callbackURI should not be used until confirmed
```suggestion
		return nil, oauthError(oauth.InvalidRequest, "invalid authorization code")
```